### DISCUSSION TITLE
How do I use slots with a Quasar Dialog Plugin Custom Component?

### DISCUSSION THREAD

**Discussion by BenJackGill** [quasarframework/quasar#12269] [General - Components / Directives / etc] [answered]
I want to make a custom component for the Quasar Dialog. And inside that component I want to use slots, but I'm not sure how to do that.

This is my `CustomDialogComponent.vue` where I have defined a `cancelBtn` slot and a `confirmBtn` slot:

    <template>
      
      <q-dialog ref="dialogRef" @hide="onDialogHide">
        <q-card class="q-dialog-plugin">
          <q-card-section>
            <strong>{{ title }}</strong>
          </q-card-section>
          <q-card-section>
            <slot name="cancelBtn" @click="handleCancelClick"></slot>
            <slot name="confirmBtn" @click="handleConfirmClick"></slot>
          </q-card-section>
        </q-card>
      </q-dialog>
    </template>
    
    <script setup lang="ts">
    import { PropType } from 'vue';
    import { useDialogPluginComponent } from 'quasar';
    
    defineProps({
      title: {
        type: String,
        required: false,
        default: 'Alert',
      },
    });
    
    defineEmits([
      ...useDialogPluginComponent.emits,
    ]);
    
    const { dialogRef, onDialogHide, onDialogOK, onDialogCancel } =
      useDialogPluginComponent();
    
    const handleConfirmClick = () => {
      console.log('Confirm Button Clicked');
      onDialogOK();
    };
    
    const handleCancelClick = () => {
      console.log('Cancel Button Clicked');
      onDialogCancel();
    };
    </script>

And the [Quasar docs][1] show that I can invoke it via a `$q.dialog({ ... })` Object. With props etc all set inside that object. So that would look something like this:

    <template>
      <div @click="showDialog">Show The Dialog</div>
    </template>
    
    <script setup lang="ts">
    import { useQuasar } from 'quasar';
    import CustomDialogComponent from 'src/components/CustomDialogComponent.vue'
    
    const $q = useQuasar();
    
    const showDialog = () => {
      $q.dialog({
        component: CustomDialogComponent,
    
        // props forwarded to your custom component
        componentProps: {
          title: 'Alert title goes here',
        },
      })
    };
    </script>

But there are no properties inside the Dialog Object for  me to pass in my slots. So where can I pass in the `cancelBtn` and `confirmBtn` slots I created in `CustomDialogComponent.vue`?

  [1]: https://quasar.dev/quasar-plugins/dialog#introduction

**Comment by yusufkandemir** [comment, marked as answer]
It's not possible to pass slots to custom dialog components. Even if it was possible, you would need to use [render functions](https://v3.vuejs.org/guide/render-function.html#render-functions) to achieve that. If there is a legit use case, and it's technically feasible, we might consider allowing passing slots using render functions.
Would you mind explaining your use case? What would you pass to `cancelBtn` and `confirmBtn` slots, is using props not enough, etc.?

**Reply by BenJackGill** [reply]
Ok thanks for letting me know.

I have created a variety of "Base" buttons which are wrapped `q-btn`'s with added custom props and styles.

I wanted to pass in those buttons as the submit and confirm buttons to the dialog component as slots. This would allow me to choose the base button by passing in slots. I could hardcode them all in the dialog component and active them using a prop and v-if, but that seemed messy and verbose.

**Reply by BenJackGill** [reply]
Update: In the end I just refactored my base buttons, which worked out better for me. Maybe there are other use cases for slots in the plugin though.

**Reply by david-mohr** [reply]
@yusufkandemir I have a similar desire to use slots in dialogs. My goal is to have a generic dialog with a variable body without the risk of using `html: true` since it will contain user input. I can pass the component and data as props and use `<component :is="bodyComponent" v-bind="bodyData">` but a slot would also be convenient. Just trying to cut down on copy/pasting the same dialog component just so I can modify the body a little (same header and buttons).
